src/wire/icmpv4.rs

```diff
@@ -265,13 +273,71 @@ impl<T: AsRef<[u8]>> Packet<T> {
         NetworkEndian::read_u16(&data[field::ECHO_SEQNO])
     }
 
+    /// Return the identifier field (for timestamp request and reply packets).
+    ///
+    /// # Panics
+    /// This function may panic if this packet is not an timestamp request or reply
+    /// packet.
+    #[inline]
+    pub fn timestamp_ident(&self) -> u16 {
+        let data = self.buffer.as_ref();
+        NetworkEndian::read_u16(&data[field::TM_IDENT])
+    }
+
+    /// Return the sequence number field (for timestamp request and reply packets).
+    ///
+    /// # Panics
+    /// This function may panic if this packet is not an timestamp request or reply
+    /// packet.
+    #[inline]
+    pub fn timestamp_seq_no(&self) -> u16 {
+        let data = self.buffer.as_ref();
+        NetworkEndian::read_u16(&data[field::TM_SEQNO])
+    }
+
+    /// Return the originate timestamp field (for timestamp request and reply packets).
+    ///
+    /// # Panics
+    /// This function may panic if this packet is not an timestamp request or reply
+    /// packet.
+    #[inline]
+    pub fn timestamp_origin_tm(&self) -> u32 {
+        let data = self.buffer.as_ref();
+        NetworkEndian::read_u32(&data[field::TM_ORIGIN_TM])
+    }
+
+
+    /// Return the receive timestamp field (for timestamp request and reply packets).
+    ///
+    /// # Panics
+    /// This function may panic if this packet is not an timestamp request or reply
+    /// packet.
+    #[inline]
+    pub fn timestamp_rx_tm(&self) -> u32 {
+        let data = self.buffer.as_ref();
+        NetworkEndian::read_u32(&data[field::TM_RX_TM])
+    }
+
+    /// Return the transmit timestamp field (for timestamp request and reply packets).
+    ///
+    /// # Panics
+    /// This function may panic if this packet is not an timestamp request or reply
+    /// packet.
+    #[inline]
+    pub fn timestamp_tx_tm(&self) -> u32 {
+        let data = self.buffer.as_ref();
+        NetworkEndian::read_u32(&data[field::TM_TX_TM])
+    }
+
     /// Return the header length.
     /// The result depends on the value of the message type field.
     pub fn header_len(&self) -> usize {
         match self.msg_type() {
-            Message::EchoRequest    => field::ECHO_SEQNO.end,
-            Message::EchoReply      => field::ECHO_SEQNO.end,
-            Message::DstUnreachable => field::UNUSED.end,
+            Message::EchoRequest      => field::ECHO_SEQNO.end,
+            Message::EchoReply        => field::ECHO_SEQNO.end,
+            Message::DstUnreachable   => field::UNUSED.end,
+            Message::TimestampRequest => field::TM_TX_TM.end,
```

Comment:
Added `TimestampRequest` and `Timestamp` reply. They're not essential to this PR. The main reason for adding them was for the test case in `iface::ethernet::test::test_icmpv4_socket`